### PR TITLE
CORE-20044 Change how idempotence is enforced on persistence requests

### PR DESCRIPTION
**Pull request by ben-millar**
### Context:
Each external event is associated with a `requestId` which is generated deterministically based on:
1. The ID of the calling flow;
2. The number of times the calling flow has suspended;
3. A hash of the parameters of the object being sent.

This `requestId` is used for de-duplication of replayed requests at the side of the external event receiver. However, we observed issues in repeated persistence requests where the entities to be persisted contained some non-deterministic field (such as a timestamp) which could change between event replays.

### Solution:
Currently, the `ProcessorService`, which handles persistence requests, maintains a repository of `requestIds` which is uses to detect duplicated events. This PR adds some logic within that service which uses  the `EntityManager` to retrieve the primary keys (PK) for object we're trying to persist from JPA/Hibernate. These PKs, if present, are used to create a new, truly deterministic hash which is swapped into the `requestId` (with the help of some new helper functions).

We can then use the normal means of deduplication.

### CODE REVIEW

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/executor/ExternalEventExecutorImpl.kt [hunk 1]
```diff
@@ -27,15 +28,13 @@ class ExternalEventExecutorImpl @Activate constructor(
...
-        val uuid = deterministicBytesID(parameters)
-
         @Suppress("unchecked_cast")
         return with(flowFiberService.getExecutingFiber()) {
             suspend(
+                // `requestId` is a deterministic ID per event which allows us to achieve idempotency by de-duplicating events processing;
+                //  A deterministic ID is required so that events replayed from the flow engine won't be reprocessed on the consumer-side.
                 FlowIORequest.ExternalEvent(
-                    generateRequestId(uuid, this),
+                    RequestID.generateRequestId(parameters, this, serializationService),
```
Comment (rick-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/executor/ExternalEventExecutorImpl.kt]: I wonder if we should generate this up where the old `uuid` was created, so it is outside of the suspend?  I don't think it matters but worth a thought.

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/PersistenceServiceInternal.kt [hunk 2]
```diff
@@ -53,17 +54,23 @@ class PersistenceServiceInternal(
...
 
-
     fun persist(
         serializationService: SerializationService,
         entityManager: EntityManager,
         payload: PersistEntities
     ): EntityResponse {
-        payload.entities.map { entityManager.persist(serializationService.deserialize(it.array(), Any::class.java)) }
+        // Don't persist the first time
+        if (callCount.incrementAndGet() > 1) {
```
Comment (rick-r3) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/PersistenceServiceInternal.kt]: I assume this is you trying to get the test doing something?

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt [hunk 3]
```diff
@@ -189,4 +198,15 @@ class ProcessorService {
         }
     }
 
-}
\ No newline at end of file
+    private fun replaceHash(
```
Comment (rick-r3) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt]: As you already know, we want to consolidate this logic _somewhere_.  Think you mentioned this code doesn't know about flows so the shared code cannot know about flows/fibers.  Maybe make `generateRequestId()` in `RequestID` take suspend count and flow ID as params, rather than a flow/fiber.  Then it is free of the flow dependency and can be put somewhere both places we need to use it can reach.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/executor/ExternalEventExecutorImpl.kt [hunk 4]
```diff
@@ -52,20 +51,33 @@ class ExternalEventExecutorImpl @Activate constructor(
...
-    }
 
-    private fun hash(bytes: ByteArray) = toBase64(bytes.sha256Bytes())
+    object RequestID {
+        // - is not a Base64 character
+        private const val DELIMITER = "-"
+
+        fun generateRequestId(
+            deterministic: Any,
+            flowFiber: FlowFiber,
```
Comment (rick-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/executor/ExternalEventExecutorImpl.kt]: Maybe we make this take suspend count and flow ID as params rather than flow fibre, then this code is not dependent on flow code and can be put somewhere it can be shared.

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/PersistenceServiceInternal.kt (hunk 2 above)
Comment (ben-millar) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/PersistenceServiceInternal.kt]: That's right; I wanted it to skip over the actual persist the first time so that we fail only on the RequestID collision. (Actually, this was probably unnecessary as we hit the requestID collision before we attempt the persist anyway)

**Comment by ben-millar** [comment]
build e2e

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        serializationService: SerializationService,
+        hashFunction: (ByteArray) -> String
+    ): String {
+        val hashedInput = deterministicBytesID(serializationService, parameters, hashFunction)
+
+        return listOf(flowId, suspendCount, hashedInput)
+            .joinToString(DELIMITER)
+    }
+
+    fun replaceHash(
```
Comment (kyriathar) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: nit: maybe rename to `replaceRequestId`, hash really is implementation detail of the id.

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package net.corda.flow.utils
+
+import net.corda.v5.application.serialization.SerializationService
+
+/**
+ * This class allows us to generate a unique request ID which is deterministic based on the parameters passed.
+ *
+ * This is used to ensure that events replayed from the flow engine won't be reprocessed on the consumer side.
+ */
+object RequestID {
```
Comment (kyriathar) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: nit: maybe rename to `RequestIdGenerator`?

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    // - is not a Base64 character
+    private const val DELIMITER = "-"
+
+    fun generateRequestId(
+        parameters: Any,
```
Comment (kyriathar) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: This `parameters` what kind of parameters is it? Or if they are generally objects that will go through `serializationService` then maybe just `objects` or something

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/PersistenceServiceInternal.kt [hunk 8]
```diff
@@ -59,11 +58,10 @@ class PersistenceServiceInternal(
 
 
     fun persist(
-        serializationService: SerializationService,
         entityManager: EntityManager,
-        payload: PersistEntities
+        payload: List<Any>
```
Comment (kyriathar) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/PersistenceServiceInternal.kt]: Maybe this could be called `entities` now to be a bit more descriptive?

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    ): String {
+        val hashedInput = hashFunction(serializationService.serialize(deterministic).bytes)
+        return requestId.substringBeforeLast(DELIMITER) + DELIMITER + hashedInput
+    }
+
+    private fun deterministicBytesID(
+        serializationService: SerializationService,
+        parameters: Any,
+        hashFunction: (ByteArray) -> String
+    ) = hashFunction(serializationService.serialize(parameters).bytes)
```
Comment (kyriathar) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: I think it might be that if parameters include somewhere some collection? If that's the case then I think that collection's traversal will need to be deterministic for this to work. Otherwise it could break. For example two sets might be equal but have different traversal orders based on population orders. Then I think this could produce a different hash for two equal sets. But not sure if this can happen here.

**Review by kyriathar — COMMENTED** [review]
A few comments

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt (hunk 5 above)
Comment (ben-millar) [reply libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: I see what you're saying. We're not replacing the entire request ID, though, only the deterministic hash at the end.

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt [hunk 10]
```diff
@@ -103,11 +106,24 @@ class ProcessorService {
...
         return when (val entityRequest = request.request) {
             is PersistEntities -> {
-                val requestId = request.flowExternalEventContext.requestId
+                val originalRequestId = request.flowExternalEventContext.requestId
+                val entities = entityRequest.entities
+                    .map { serializationService.deserialize(it.array(), Any::class.java) }
+
+                val requestId = entities
+                    .mapNotNull { entityManagerFactory.persistenceUnitUtil.getIdentifier(it) }
+                    .takeIf { it.isNotEmpty() }?.let { entityPks ->
```
Comment (driessamyn) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt:123]: This doesn't guarantee idempotency, but instead I think this just replicates the `merge` behaviour.
What I mean is that you can send a different payload and it will not behave you expected I think.

For example, image naive (pseudo)flow code that does:

```
persistenceService.persist("dog-table", "Felix", Instance.now())
persistenceService.persist("dog-table", "Felix", Instance.now())
```

Where "Felix" is the value for the "dog-table" primary key.
You would expect the second operation to throw a key constraint exception, but I think this code changes it to the same behaviour as merge.

Comment (ben-millar) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt:123]: Good question, but I don't think that's correct. The first thing that will stop that from happening is that the `suspendCount` part of the `requestId` will be incremented, and so the second call will not be de-duplicated and will fail.

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt (hunk 5 above)
Comment (kyriathar) [reply libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: Then perhaps we could rename to somehow describe that?

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt (hunk 10 above)
Comment (ben-millar) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt:123]: If the flow code was simply:
```
persistenceService.persist("dog-table", "Felix", Instance.now())
```
But something went wrong such that the `persist` happened but wasn't reported back to the flow engine, and the flow engined retried the event:

**Existing code:** 
If the timestamp `Instance.now()` has changed since the first request, it will result in a new, unique `requestId` and the event will **not** be de-duplicated by the persistence worker.

**New code:**
Although the `requestId` has been changed due to the event having some non-deterministic parameter (the timestamp), we will instead use the JPA ID of the entity to check for duplicates, and the persistence worker will successfully de-duplicate the request. 

I've tested this scenario, and the request IDs for each were:

**Original**
Request 1: bf13ea58-63ff-4562-b664-14af5ca7fcb4-1-5ewQjaxE4l3jlcxkZgLS24Sp8d9j6KfrPjT7Xrg/iUI=
Request 2: bf13ea58-63ff-4562-b664-14af5ca7fcb4-1-45FtZGvEidUdWnKk2dTBPqz35F8AB0f02xEFQ7nx4Js=

**"New" RequestID**
Request 1: bf13ea58-63ff-4562-b664-14af5ca7fcb4-1-4yAETfhRP793Uf1oZ2d7ROgYdgglONrEFbYdIDHCeYI=
Request 2: bf13ea58-63ff-4562-b664-14af5ca7fcb4-1-4yAETfhRP793Uf1oZ2d7ROgYdgglONrEFbYdIDHCeYI=

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package net.corda.flow.utils
+
+import net.corda.v5.application.serialization.SerializationService
+
+/**
+ * This class allows us to generate a unique request ID which is deterministic based on the parameters passed.
```
Comment (simon-johnson-r3) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: This isn't a class

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * This is used to ensure that events replayed from the flow engine won't be reprocessed on the consumer side.
+ */
+object RequestID {
+    // - is not a Base64 character
```
Comment (simon-johnson-r3) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: minor, but I'd change this to `hyphen is not a Base64 character` for readability

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt (hunk 7 above)
Comment (simon-johnson-r3) [reply libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: I think maybe `requestParameters` would be more expressive? I think that's what they are, and perhaps document the function too.

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        requestId: String,
+        deterministic: Any,
+        serializationService: SerializationService,
+        hashFunction: (ByteArray) -> String
+    ): String {
+        val hashedInput = hashFunction(serializationService.serialize(deterministic).bytes)
+        return requestId.substringBeforeLast(DELIMITER) + DELIMITER + hashedInput
+    }
+
+    private fun deterministicBytesID(
```
Comment (simon-johnson-r3) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: I find the existence of this one line function obfuscates the code a bit. I expect it's evolved like this because a similar named function existed already. But it's only called in one place, and it requires quite a few parameters to be passed in, including a function, the call of which is basically more or less the only thing the function does.

I would remove it.

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt (hunk 9 above)
Comment (simon-johnson-r3) [reply libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestID.kt]: That could be a problem but I'd expect the container type to handle it appropriately. For example for a `Set` the hashcodes are added together 

https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/util/Set.html#hashCode()

`Returns the hash code value for this set. The hash code of a set is defined to be the sum of the hash codes of the elements in the set, where the hash code of a null element is defined to be zero. This ensures that s1.equals(s2) implies that s1.hashCode()==s2.hashCode() for any two sets s1 and s2, as required by the general contract of [Object.hashCode()](https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/lang/Object.html#hashCode()).`

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestIDGenerator.kt [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+ * was generated. This improvement is only possible at the consumer side of the request, and also only possible under
+ * circumstances where we can extract a unique deterministic entity id from the entities in the request. Where an entity
+ * id is available it is more reliable that the request parameters, because it's the source of uniqueness enforced at the
+ * db level already. This is particularly important where parameters passed to the request are nondeterministic which typically
+ * occurs where one of them is a timestamp.
+ *
+ * The format of the request ID is <flowId>-<suspendCount>-<entity_based_hash> where entity_based_hash is either based on
+ * the parameters to the request or the entity's known unique id.
+ */
+object RequestIDGenerator {
```
Comment (kyriathar) [inline libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestIDGenerator.kt:20]: nit: looking at the documentation of this class it seems it is meant to be for flow request ids? So maybe we could rename to `FlowRequestIdGenerator` or  similar?

**Review by rick-r3 — APPROVED** [review]

File: libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestIDGenerator.kt (hunk 14 above)
Comment (simon-johnson-r3) [reply libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/RequestIDGenerator.kt:20]: Discussed offline, this is already in a flow package, so we agreed not to repeat the word "flow".

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt (hunk 10 above)
Comment (simon-johnson-r3) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt:123]: I've documented the use of suspend count to alleviate this issue in the form of comments in the code.

Comment (driessamyn) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/ProcessorService.kt:123]: I still don't think this answers the question.

- If we increase the suspension count for each persistence request, then I don't see why we need to "recalculate the request ID" based on PK as the flow ID and suspension count should be enough. I.e. why do we need the parameters?
- If we don't increase the suspension count for each persistence request, then we have a problem as we would silently dedupe genuine requests that should error instead.

In short I don't see how a hash of the key fields helps us.
